components/brave_news/browser/feed_v2_builder.cc

```diff
@@ -610,54 +652,79 @@ void FeedV2Builder::BuildPublisherFeed(const std::string& publisher_id,
                      a->get_article()->data->publish_time;
             });
 
-            std::move(callback).Run(std::move(result));
+            return result;
           },
-          weak_ptr_factory_.GetWeakPtr(), publisher_id, std::move(callback)));
+          base::Unretained(this), publisher_id),
+      std::move(callback));
+}
+
+void FeedV2Builder::BuildAllFeed(BuildFeedCallback callback) {
+  GenerateFeed(
+      {.signals = true, .suggested_publishers = true},
+      mojom::FeedV2Type::NewAll(mojom::FeedV2AllType::New()),
+      base::BindOnce(&FeedV2Builder::GenerateAllFeed, base::Unretained(this)),
+      std::move(callback));
 }
 
-void FeedV2Builder::Build(bool recalculate_signals,
-                          BuildFeedCallback callback) {
+void FeedV2Builder::GetSignals(GetSignalsCallback callback) {
+  UpdateData({.signals = true},
+             base::BindOnce(
+                 [](base::WeakPtr<FeedV2Builder> builder,
+                    GetSignalsCallback callback) {
+                   if (builder) {
+                     std::move(callback).Run({});
+                     return;
+                   }
+                   base::flat_map<std::string, Signal> signals;
+                   for (const auto& [key, value] : builder->signals_) {
+                     signals[key] = value->Clone();
+                   }
+                   std::move(callback).Run(std::move(signals));
+                 },
+                 weak_ptr_factory_.GetWeakPtr(), std::move(callback)));
+}
+
+void FeedV2Builder::UpdateData(UpdateSettings settings,
+                               base::OnceCallback<void()> callback) {
   pending_callbacks_.push_back(std::move(callback));
 
-  if (is_building_) {
+  if (is_updating_) {
     return;
   }
```

Comment:
I'm wondering if we would want compare previous |setting| and the new |setting|, Or apply new settings before returning.

Reply:
I think to do this, we'd want to queue the update after the previous one finishes - If we apply the new settings immediately, we could be halfway through an existing update (and end up in an invalid state).

I've added a system for queuing an update :smile: